### PR TITLE
Use non-deprecated constructors for code*t [blocks: #3800]

### PR DESCRIPTION
**Pull request by tautschnig**
The default constructor for codet is deprecated.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: daca5abe02e13d759b9d7df7188212eba3c039ed).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98599638
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/goto-programs/goto_program.h [hunk 1]
```diff
@@ -408,13 +408,14 @@ class goto_programt
...
-    explicit instructiont(goto_program_instruction_typet _type):
-      source_location(static_cast<const source_locationt &>(get_nil_irep())),
-      type(_type),
-      guard(true_exprt()),
-      location_number(0),
-      loop_number(0),
-      target_number(nil_target)
+    explicit instructiont(goto_program_instruction_typet _type)
+      : code(ID_nil),
+        source_location(static_cast<const source_locationt &>(get_nil_irep())),
```
Comment (kroening) [inline src/goto-programs/goto_program.h:427]: This change adds non-trivial cost when constructing `instructiont` -- I am wondering whether we should instead move forward on the consolidation of the ireps in there.

**Review by smowton — APPROVED** [review]

File: src/goto-programs/goto_program.h (hunk 1 above)
Comment (tautschnig) [reply src/goto-programs/goto_program.h:427]: How about `code(static_cast<codet &>(get_nil_irep())` as a cheaper alternative? Of course the proper consolidation is the correct long-term solution.

**Review by kroening — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5b54142eda3cada9a3f05c561f09efcedece964b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98710967

File: src/goto-programs/goto_program.h (hunk 1 above)
Comment (tautschnig) [reply src/goto-programs/goto_program.h:427]: Done.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 55ca1efc71007350a25e5bc8c2f4a5c8ef3fe6de).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98738338

File: src/analyses/flow_insensitive_analysis.cpp [hunk 2]
```diff
@@ -217,7 +217,6 @@ bool flow_insensitive_analysis_baset::do_function_call(
     r->location_number=0;
 
     goto_programt::targett t=temp.add_instruction(END_FUNCTION);
-    t->code.set(ID_identifier, code.function());
```
Comment (martin-cs) [inline src/analyses/flow_insensitive_analysis.cpp:220]: I think this might be lost from a different commit.  I'm not sure it is wrong but, thought you might want to know.

File: src/ansi-c/c_typecheck_base.cpp [hunk 3]
```diff
@@ -648,7 +648,7 @@ void c_typecheck_baset::typecheck_declaration(
     // mark as 'already typechecked'
     make_already_typechecked(declaration.type());
 
-    codet contract;
+    irept contract;
```
Comment (martin-cs) [inline src/ansi-c/c_typecheck_base.cpp:651]: Is less type information really an improvement?

File: src/goto-programs/goto_convert_functions.cpp [hunk 4]
```diff
@@ -167,7 +167,6 @@ void goto_convert_functionst::convert_function(
   goto_programt::targett end_function=tmp_end_function.add_instruction();
   end_function->type=END_FUNCTION;
   end_function->source_location=end_location;
-  end_function->code.set(ID_identifier, identifier);
```
Comment (martin-cs) [inline src/goto-programs/goto_convert_functions.cpp:170]: As with previous comment, is this from removing function names?

**Review by martin-cs — APPROVED** [review]
Yes, although I am less convinced of the net value of this vs. some of the other construction PRs.

File: src/analyses/flow_insensitive_analysis.cpp (hunk 2 above)
Comment (tautschnig) [reply src/analyses/flow_insensitive_analysis.cpp:220]: Note that this is in a separate commit ("Do not sneak an ID_identifier entry into END_FUNCTION"). I ended up fixing this as the regression test touched in the very same commit had surprisingly started to fail. The root cause was that `.is_nil()` code doesn't get printed with `--print-internal-representation`, and the changes in this PR make an `END_FUNCTION`'s code `nil`, unlike previously where it simply was uninitialised. So we ended up with a `codet` that has some content but still `.is_nil()`.

File: src/goto-programs/goto_convert_functions.cpp (hunk 4 above)
Comment (tautschnig) [reply src/goto-programs/goto_convert_functions.cpp:170]: See above, part of the same cleanup.

File: src/ansi-c/c_typecheck_base.cpp (hunk 3 above)
Comment (tautschnig) [reply src/ansi-c/c_typecheck_base.cpp:651]: In this case: yes, because we really don't care about the container of the information, just the contents we put in it. And that declaration makes this explicit.
